### PR TITLE
fix: use the native `fetch` API in browsers and a patched `node-fetch` otherwise

### PR DESCRIPTION
**Pull request by steveluscher**
#### Problem

1. Presently, web3.js does not work in a Service Worker. `cross-fetch` is incompatible in that environment.
2. We're including a `fetch` ponyfill when in reality web3.js is already incompatible with browsers that don't support `fetch` natively.

#### Summary of Changes

* Augment the build system so that you can fork modules
* Fork the browser version of the fetch implementation to very simply be `globalThis.fetch` which works in browser and Service Worker environments.

#### Notes on the new fork system

The fork system (which should really have been in a separate PR) lets you override modules depending on the build environment. Given that you have a module called `foo`, the build system will check if there's a module at that directory level called `__forks__/CONFIG/foo.ts`. If there is, that will be built into the bundle.

```
src/
L __forks__/
  L browser/
    L foo.ts  <-- this will be used in browser builds
L foo.ts  <-- this will be used in non-browser builds
L utils/
  L __forks__/
    L node/
      L assert.ts  <-- this will be used in node builds
  L assert.ts  <-- this will be used in non-node builds
``` 

#### Test plan

Ran the test plan at https://github.com/wentokay/unfetch-test. Thanks @wentokay!

Noted that this saves 2.62K gzipped (tested in Solana Explorer build).

---
This diff was inspired by the phenomenal PR #25500.

### CODE REVIEW

File: web3.js/src/__forks__/browser/fetch-impl.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+export default globalThis.fetch;
```
Comment (steveluscher) [inline web3.js/src/__forks__/browser/fetch-impl.ts:1]: `globalThis` will resolve to `self` in workers, and `window` in browsers. Both have `fetch` available on them.

**Comment by jstarry** [comment]
Aside.. @steveluscher what version of yarn do you use? I'm on 1.22.18 and I think we keep thrashing that file and creating huge diffs. Can we try to get on the same version?

**Comment by steveluscher** [comment]
I'm on 1.22.17. 🤔

**Comment by jstarry** [comment]
Maybe it's dependabot then 🤷🏻

**Comment by steveluscher** [comment]
That little point version made all the difference! I just upgraded to 1.22.19 and now the diff is many times smaller.

File: web3.js/src/fetch-impl.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import * as nodeFetch from 'node-fetch';
+
+export * from 'node-fetch';
+export default async function (
+  input: nodeFetch.RequestInfo,
+  init?: nodeFetch.RequestInit,
+): Promise<nodeFetch.Response> {
+  const processedInput =
+    typeof input === 'string' && input.slice(0, 2) === '//'
```
Comment (jstarry) [inline web3.js/src/fetch-impl.ts:9]: when would the input be missing a protocol and just start with `//`?

File: web3.js/src/connection.ts [hunk 3]
```diff
@@ -955,23 +955,21 @@ function createRpcClient(
   url: string,
   useHttps: boolean,
   httpHeaders?: HttpHeaders,
-  customFetch?: typeof crossFetch,
+  customFetch?: FetchFn,
   fetchMiddleware?: FetchMiddleware,
   disableRetryOnRateLimit?: boolean,
 ): RpcClient {
-  const fetch = customFetch ? customFetch : crossFetch;
+  const fetchImpl = (customFetch ? customFetch : fetch) as FetchFn;
```
Comment (jstarry) [inline web3.js/src/connection.ts:962]: nit: how about importing `./fetch-iml` as `fetchImpl` and keeping this as `fetch`?

File: web3.js/src/connection.ts [hunk 4]
```diff
@@ -2162,14 +2160,26 @@ export type ConfirmedSignatureInfo = {
  */
 export type HttpHeaders = {[header: string]: string};
 
+export type AlterReturnType<TFn, TReturn> = TFn extends (
+  ...args: infer A
+) => any
+  ? (...args: A) => TReturn
+  : never;
```
Comment (jstarry) [inline web3.js/src/connection.ts:2167]: I can't really make sense of this syntax.. what is being evaluated by the `?` operator?

File: web3.js/src/connection.ts [hunk 5]
```diff
@@ -2162,14 +2160,26 @@ export type ConfirmedSignatureInfo = {
...
  */
-export type FetchMiddleware = (
-  url: string,
-  options: any,
-  fetch: (modifiedUrl: string, modifiedOptions: any) => void,
-) => void;
+export type FetchMiddleware = AlterReturnType<
+  AddParameters<FetchFn, [fetch: AlterReturnType<FetchFn, void>]>,
+  void
+>;
```
Comment (jstarry) [inline web3.js/src/connection.ts:2182]: What is the benefit of expressing the type of `FetchMiddleware` like this over the original type signature which is more explicit?

**Review by jstarry — COMMENTED** [review]
lgtm but have a few questions.. and I think the typing magic you used is cool but hard to understand and maintain. I'd prefer if things were simpler.

File: web3.js/src/fetch-impl.ts (hunk 2 above)
Comment (steveluscher) [reply web3.js/src/fetch-impl.ts:9]: This is a standard patch for the broken implementation of `node-fetch` that doesn't support protocol-relative URLs like the `fetch` in browsers does. https://github.com/node-fetch/node-fetch/blob/main/docs/v2-LIMITS.md

Folks can and will write apps that make use of protocol-relative URLs, and we don't want them to break in Node when they do.

This replicates what `cross-fetch` was already doing.

https://github.com/lquixada/cross-fetch/blob/main/src/node-ponyfill.js#L4-L11

File: web3.js/src/connection.ts (hunk 4 above)
Comment (steveluscher) [reply web3.js/src/connection.ts:2167]: Fair enough. I was hoping the names of the utility types would make the reader incurious as to how it's implemented. These types should really just be TypeScript builtins, but alas.

I'll kill this stuff and just write the types manually.

File: web3.js/src/connection.ts (hunk 5 above)
Comment (steveluscher) [reply web3.js/src/connection.ts:2182]: Well, the original signature is just straight up wrong. The first parameter to `fetch` is `RequestInfo` not `string`. The second `RequestInit` not `any`. I'm just trying to make everything properly line up with the type of the `fetch-impl` shim.

**Pull request merged**